Implement a lazy tiff loader

tjamet opened the pull request:
Implement a lazy tiff loader

When using tiff files like raw files, the time required to load the
whole file is significant. We do not load the whole picture frame in-memory
in order to access EXIF values. We do not even need to load all
EXIF values when we need to access a single EXIF field.

To improve the tiff loading speed, opportunistically read from the
source file when needed.

This is a significant change in the interface as it requires to pass a
ReaderAt instead of a Reader.

In order not to break backward compatibility, a LazyDecode function has
been implemented to keep the same interface for the Decode one,
including filling every Tag.

There are 2 places where the backward compatibility is broken:
- DecodeTag now requires a ReaderAt and the base offset of the tag
- In case of using an open file to feed the tiff decoder, the file needs
  to stay open during the whole life cycle of the tiff structure

Benchmark results (on SSD):

EXIF parsing

```
goos: darwin
goarch: amd64
pkg: github.com/rwcarlsen/goexif/exif
BenchmarkDecode-8                 200000             89622 ns/op          133531 B/op        236 allocs/op
BenchmarkDecodeRaw-8               30000            486907 ns/op          341765 B/op       8553 allocs/op
BenchmarkLazyDecode-8             200000             83383 ns/op           31659 B/op        151 allocs/op
BenchmarkLazyDecodeRaw-8          200000             81325 ns/op           25051 B/op        164 allocs/op
PASS
ok      github.com/rwcarlsen/goexif/exif        72.727s
```

Before update, the benchmark results for exif parsing were:

```
goos: darwin
goarch: amd64
pkg: github.com/rwcarlsen/goexif/exif
BenchmarkDecode-8         100000            170564 ns/op          342413 B/op        961 allocs/op
BenchmarkDecodeRaw-8         100         112034369 ns/op        750784604 B/op      9244 allocs/op
PASS
ok      github.com/rwcarlsen/goexif/exif        34.500s
```

TIFF decoding

```
goos: darwin
goarch: amd64
pkg: github.com/rwcarlsen/goexif/tiff
BenchmarkDecode-8                    300          46751365 ns/op        268452505 B/op       113 allocs/op
BenchmarkLazyDecode-8             300000             43886 ns/op           10645 B/op         82 allocs/op
PASS
ok      github.com/rwcarlsen/goexif/tiff        32.337s
```

tjamet committed:
Add benchmark test

Here are the results ran locally on a SSD:

goos: darwin
goarch: amd64
pkg: github.com/rwcarlsen/goexif/tiff
BenchmarkDecode-8       Skip downloading existing raw file http://www.rawsamples.ch/raws/canon/RAW_CANON_EOS_5DS.CR2
     200          43701603 ns/op        268632324 B/op      8709 allocs/op
PASS
ok      github.com/rwcarlsen/goexif/tiff        10.294s

Signed-off-by: Thibault Jamet <tjamet@users.noreply.github.com>

tjamet committed:
Implement a lazy loader

When using tiff files like raw files, the time required to load the
whole file is significant. We do not load the whole picture frame in-memory
in order to access EXIF values. We do not even need to load all
EXIF values when we need to access a single EXIF field.

To improve the tiff loading speed, opportustically read from the
source file when needed.

This is a significant change in the interface as it requires to pass a
ReaderAt instead of a Reader.

In order not to break backward compatibility, a LazyDecode function has
been implemented to keep the same interface for the Decode one,
including filling every Tag.

There are 2 places where the backward compatibility is broken:
- DecodeTag now requires a ReaderAt and the base offset of the tag
- In case of using an open file to feed the tiff decoder, the file needs
  to stay open during the whole life cycle of the tiff structure

Benchmark results (on SSD):

EXIF parsing

goos: darwin
goarch: amd64
pkg: github.com/rwcarlsen/goexif/exif
BenchmarkDecode-8                 200000             89622 ns/op          133531 B/op        236 allocs/op
BenchmarkDecodeRaw-8               30000            486907 ns/op          341765 B/op       8553 allocs/op
BenchmarkLazyDecode-8             200000             83383 ns/op           31659 B/op        151 allocs/op
BenchmarkLazyDecodeRaw-8          200000             81325 ns/op           25051 B/op        164 allocs/op
PASS
ok      github.com/rwcarlsen/goexif/exif        72.727s

Before update, the benchmark results for exif parsing were:

goos: darwin
goarch: amd64
pkg: github.com/rwcarlsen/goexif/exif
BenchmarkDecode-8         100000            170961 ns/op          342412 B/op        961 allocs/op
BenchmarkDecodeRaw-8         100         109398876 ns/op        750784550 B/op      9244 allocs/op
PASS
ok      github.com/rwcarlsen/goexif/exif        33.940s

TIFF decoding

goos: darwin
goarch: amd64
pkg: github.com/rwcarlsen/goexif/tiff
BenchmarkDecode-8                    300          46751365 ns/op        268452505 B/op       113 allocs/op
BenchmarkLazyDecode-8             300000             43886 ns/op           10645 B/op         82 allocs/op
PASS
ok      github.com/rwcarlsen/goexif/tiff        32.337s

Signed-off-by: Thibault Jamet <tjamet@users.noreply.github.com>

tjamet committed:
Add test to verify that Decode loads tag values

Future commits will use lazy load for values.
For backward compatibility, Decode will keep loading all tag values.

Signed-off-by: Thibault Jamet <tjamet@users.noreply.github.com>

rwcarlsen commented on exif/exif.go:
```diff
@@ -209,92 +210,45 @@ type Exif struct {
 // The error can be inspected with functions such as IsCriticalError
 // to determine whether the returned object might still be usable.
-func Decode(r io.Reader) (*Exif, error) {
+func Decode(r io.ReaderAt) (*Exif, error) {
```
I don't think I'm okay with this significant breakage of the API.

tjamet replied:
Would the same principle as in the TIFF decoder (`Decode` and `LazyDecode`) be a satisfying alternative for you?

rwcarlsen replied:
That would certainly help - that would leave the only other API breakage being the removal of the exported Exif.Raw field.  Perhaps we could leave Raw in the struct, but only populate it when a doing a non-lazy decode - that should allow existing API behavior to be mostly preserved.

rwcarlsen reviewed (changes requested):
goexif has been around long enough, I don't think we should be making big API changes like this.  If you can find a way to avoid the significant API changes I'd be happy to work to get things in.  Otherwise, it might be time to start working toward a v2 - especially since the vgo/modules proposal is being adopted into go1.11

tjamet committed:
Add a test to ensure exif.Raw contains a valid exif file

Signed-off-by: Thibault Jamet <tjamet@users.noreply.github.com>

tjamet commented:
I have changed the implementation so that the interface is kept, and added a test to actually ensure that, after `Decode`, the `Raw` field has a valid tiff structure

rwcarlsen commented:
Thanks! I'll try and take a look sometime this week.

rwcarlsen committed:
tweaks
